src/couch_replicator_doc_processor.erl

```diff
@@ -54,89 +83,269 @@ db_change(DbName, {ChangeProps} = Change, Server) ->
     Rep = Rep0#rep{db_name = DbName},
-    case couch_replicator_scheduler:rep_state(RepId) of
-    nil ->
-        case couch_replicator_scheduler:add_job(Rep) of
-        ok ->
-            ok;
-        {error, already_added} ->
-            couch_log:warning("replicator scheduler: ~p was already added", [Rep])
-        end,
-        ok;
-    #rep{doc_id = DocId} ->
-        ok;
-    #rep{doc_id = null} ->
-        couch_log:warning("Replication `~s` specified by document `~s`"
-            " already running as a transient replication, started via"
-            " `_replicate` API endpoint", [pp_rep_id(RepId), DocId]);
-    #rep{db_name = OtherDbName, doc_id = OtherDocId} ->
-        case mem3:dbname(OtherDbName) =:= mem3:dbname(DbName) of
-        true ->
-            Msg = io_lib:format("Replication `~s` specified by document `~s`"
-                " already started, triggered by document `~s` from the same"
-                " database", [pp_rep_id(RepId), DocId, OtherDocId]),
-            couch_log:notice(Msg, []),
-            couch_replicator_docs:update_failed(DbName, DocId, Msg);
-        false ->
-            Msg = io_lib:format("Replication `~s` specified by document `~s`"
-                " already started triggered by document `~s` from a different"
-                " database", [pp_rep_id(RepId), DocId, OtherDocId]),
-            couch_log:warning(Msg, []),
-            couch_replicator_docs:update_failed(DbName, DocId, Msg)
-        end
+    Filter = case couch_replicator_filters:parse(Rep#rep.options) of
+    {ok, nil} ->
+        nil;
+    {ok, {user, _FName, _QP}} ->
+        user;
+    {ok, {view, _FName, _QP}} ->
+        view;
+    {ok, {docids, _DocIds}} ->
+        docids;
+    {ok, {mango, _Selector}} ->
+        mango;
+    {error, FilterError} ->
+        throw(FilterError)
     end,
+    gen_server:call(?MODULE, {updated, Id, Rep, Filter}, infinity).
+
+
+% Doc processor gen_server API and callbacks
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [],  []).
+
+
+init([]) ->
+    ?MODULE = ets:new(?MODULE, [named_table, {keypos, #rdoc.id}]),
+    {ok, nil}.
+
+
+terminate(_Reason, _State) ->
     ok.
 
 
--spec remove_jobs(binary(), binary()) -> ok.
-remove_jobs(DbName, DocId) ->
+handle_call({updated, Id, Rep, Filter}, _From, State) ->
+    ok = updated_doc(Id, Rep, Filter),
+    {reply, ok, State};
+
+handle_call({removed, Id}, _From, State) ->
+    ok = removed_doc(Id),
+    {reply, ok, State};
+
+handle_call({clean_up_replications, DbName}, _From, State) ->
+    ok = removed_db(DbName),
+    {reply, ok, State}.
+
+
+handle_cast(Msg, State) ->
+    {stop, {error, unexpected_message, Msg}, State}.
+
+
+handle_info({'DOWN', Ref, _, _, #doc_worker_result{id = Id, result = Res}},
+        State) ->
+    ok = worker_returned(Ref, Id, Res),
+    {noreply, State};
+
+handle_info(Msg, State) ->
+    {stop, {error, unexpected_info_message, Msg}, State}.
```

Comment:
Good point. I mainly opted for a crash because worker is expected only return `{'DOWN', Ref, _, _, #doc_worker_result{}}` so in case it somehow returns something else, thought it should crash to make it visible. If it crashes it will restart the doc processor and everything to the right of it in the supervision tree.